File: example04retrofit/src/test/java/foo/bar/example/asafretrofit/feature/fruit/StateBuilder.java
package foo.bar.example.asafretrofit.feature.fruit;

import org.mockito.ArgumentCaptor;

import java.util.ArrayList;
import java.util.List;

import co.early.fore.core.callbacks.FailureCallbackWithPayload;
import co.early.fore.core.callbacks.SuccessCallbackWithPayload;
import co.early.fore.retrofit.CallProcessor;
import foo.bar.example.asafretrofit.api.fruits.FruitPojo;
import foo.bar.example.asafretrofit.message.UserMessage;

import static org.mockito.Matchers.any;
import static org.mockito.Mockito.doAnswer;

/**
 *
 */
public class StateBuilder {

    private CallProcessor<UserMessage> mockCallProcessor;

    StateBuilder(CallProcessor<UserMessage> mockCallProcessor) {
        this.mockCallProcessor = mockCallProcessor;
    }

    StateBuilder getFruitSuccess(final FruitPojo fruitPojo) {

        List<FruitPojo> fruitList = new ArrayList<>();
        fruitList.add(fruitPojo);

        final ArgumentCaptor<SuccessCallbackWithPayload> callback = ArgumentCaptor.forClass(SuccessCallbackWithPayload.class);

        doAnswer(__ -> {
            callback.getValue().success(fruitList);
            return null;
        })

                .when(mockCallProcessor)
                .processCall(any(), any(), any(), callback.capture(), any());

        return this;
    }

    StateBuilder getFruitFail(final UserMessage userMessage) {

        final ArgumentCaptor<FailureCallbackWithPayload> callback = ArgumentCaptor.forClass(FailureCallbackWithPayload.class);

        doAnswer(__ -> {
            callback.getValue().fail(userMessage);
            return null;
        })
        .when(mockCallProcessor)
        .processCall(any(), any(), any(), any(), callback.capture());

        return this;
    }

}

File: example04retrofit-kt/src/test/java/foo/bar/example/foreretrofitcoroutine/feature/fruit/StateBuilder.kt
package foo.bar.example.foreretrofitcoroutine.feature.fruit

import co.early.fore.core.callbacks.FailureWithPayload
import co.early.fore.core.callbacks.SuccessWithPayload
import co.early.fore.retrofit.MessageProvider
import co.early.fore.retrofit.coroutine.CallProcessor
import foo.bar.example.foreretrofitcoroutine.api.fruits.FruitPojo
import foo.bar.example.foreretrofitcoroutine.message.UserMessage
import io.mockk.every
import io.mockk.slot

/**
 *
 */
class StateBuilder internal constructor(private val mockCallProcessor: CallProcessor<UserMessage>) {

    internal fun getFruitSuccess(fruitPojo: FruitPojo): StateBuilder {

        val slot = slot<co.early.fore.core.callbacks.SuccessWithPayload<List<FruitPojo>>>()

        every {
            mockCallProcessor.processCall(any(), any() as Class<MessageProvider<UserMessage>>, capture(slot), any(), any())
        } answers {
            slot.captured(listOf(fruitPojo))
        }

        return this
    }

    internal fun getFruitFail(userMessage: UserMessage): StateBuilder {

        val slot = slot<co.early.fore.core.callbacks.FailureWithPayload<UserMessage>>()

        every {
            mockCallProcessor.processCall(any(), any() as Class<MessageProvider<UserMessage>>, any<co.early.fore.core.callbacks.SuccessWithPayload<List<FruitPojo>>>(), capture(slot), any())
        } answers {
            slot.captured.invoke(userMessage)
        }

        return this
    }
}
